[BOLT] Return proper minimal alignment from BF

yota9 opened the pull request:
Currently minimal alignment of function is hardcoded to 2 bytes.
Add 2 more cases:
1. In case BF is data in code return the alignment of CI as minimal
alignment
2. For aarch64 and riscv platforms return the minimal value of 4 (added
test for aarch64)
Otherwise fallback to returning the 2 as it previously was.

yota9 commented:
@maksfb I fill like setting minimal required alignment per BF is more proper way to fix the CI alignment issue to be honest. So we would have "minimal alignment" field that is the BFs guaranteed alignment that we would have and the alignment/maxbyte alignment non guaranteed logic. What do you think? Thanks!

maksfb commented:
> @maksfb I fill like setting minimal required alignment per BF is more proper way to fix the CI alignment issue to be honest. So we would have "minimal alignment" field that is the BFs guaranteed alignment that we would have and the alignment/maxbyte alignment non guaranteed logic. What do you think? Thanks!

Sounds like we should handle CIs separately from BFs.

yota9 commented:
A few years ago I had the same feelings. Right now I'm not really sure how to re-design the code to handle them separately to be honest.

maksfb commented:
Do you anticipate different CIs to have different min alignments in the future? If not, I would rather modify the newly added `getMinAlignment()` to return a correct value based on the underlying arch and the type of object. Turning `MinAlign` into a field needs a justification.

yota9 commented:
Well, what's came in to my mind is the golang functions, that needs at least the 16 bytes alignment. But I agree that for now we can check everything and return the result from getMinAlignment() function, let me update the review later. Thanks!

yota9 committed:
[BOLT] Return proper minimal alignment from BF

Currently minimal alignment of function is hardcoded to 2 bytes.
Add 2 more cases:
1. In case BF is data in code return the alignment of CI as minimal
alignment
2. For aarch64 and riscv platforms return the minimal value of 4 (added
test for aarch64)
Otherwise fallback to returning the 2 as it previously was.

yota9 commented:
Gentle ping

yota9 commented:
@maksfb gentle ping

maksfb reviewed (approved):
LGTM. Thanks!

mtvec commented on bolt/lib/Passes/Aligner.cpp:
```diff
@@ -158,21 +158,13 @@ void AlignerPass::runOnFunctions(BinaryContext &BC) {
-      return;
-    }
+    // Minimal code alignment on AArch64 is 4
+    if (BC.isAArch64())
```
I completely missed this ping, sorry about that! Follow-up in #69837.
